[Woo POS] Coupons: Checkout - Validating coupon state

staskus opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the GitHub issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

Introduced 'Validating Coupon' state when the coupon is added to the order. This is an artificial payment state, therefore, I couldn't introduce it cleanly into the existing structure. In the end, it's similar to 'card reader disconnected' state which is handled separately in 'TotalsView'.

The two challenges were related to transitions:

### Totals Views shimmering transition

shimmeringLineView and totalsField views weren't transitioning smoothly from one another. We haven't had a state transition yet where we transitioned from Loading+Shimmering -> Loading+Totals. I had to make some adjustments to the totals fields to make the transition go smoothly.

### A gap between syncing and loaded states

- There's a gap between order.syncing and the first card payment event. Due to this gap, if we only show 'Validating Coupon' state when the order is syncing, we see 'Validating Coupons' (order.syncing) -> a momentary blink / fade out (order.loaded & paymentState == .idle) -> 'Checking Order'. 

At first, I had added additional logic to continue showing 'Validating Coupon' until the card payment state is no longer `.idle`. However, I didn't feel comfortable with this solution.

I spent some more time on it and settled on a solution to artificially prolong `.syncing` state when coupons are in the cart. 

I don't love either approach. If we don't want to introduce such a solution to our codebase we could choose:

- Keep the visual glitch (I wouldn't like that)
- Keep this custom solution
- Remove 'Validating Coupons' state from scope

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

Test cases:
- Card Reader Disconnected + Products => Loading -> Card Reader Disconnected
- Card Reader Disconnected + Products + Coupons => Validating Coupons -> Card Reader Disconnected
- Card Reader Disconnected + Products + Invalid Coupons => Validating Coupons -> Error
- Card Reader Connected + Products =>  Loading -> Checking Order
- Card Reader Connected + Products + Coupons => Validating Coupons -> Checking Order -> Proceed to payment
- Card Reader Connected + Products + Invalid Coupons => Validating Coupons -> Error

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: https://woomobilep2.wordpress.com/2024/05/06/woocommerce-mobile-quality-report-march-april/#comment-12036 -->

- Tested with iPad Air M2 11 inch iOS 18.3 device
- Unit Tests

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->


https://github.com/user-attachments/assets/4ea8677d-7397-40c6-8655-e1cb3a6c1264

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

staskus committed:
Create PointOfSaleCardPresentPaymentValidatingCouponMessageViewModel

Validating Coupon message is an artificial payment message type to show when order is loading with coupons in a cart since at the same time coupons are validated.

staskus committed:
Add shouldShowCouponValidation to TotalsViewHelper

Order validation should be shown only when there are coupons in cart and order is syncing.

Additionally, after order finishes syncing, there's a small gap between order turning into loaded state and the first card payment events. Continue showing coupon validation during this gap.

staskus committed:
Display coupon validation in TotalsView

- Show paymentView when coupon is validating
- Display PointOfSaleCardPresentPaymentInLineMessage for couponValidation

staskus committed:
Split paymentView into smaller parts for readability

staskus committed:
Display totals fields view shimmering line in an overlay for smoother transitions

shimmeringLineView and totalsField used to be presented in different views. However, it makes the transition between one state to another not as smooth.

For a smooth transition:
- Hide rather than remove shimmeringLineView and totalsField using opacity
- Display shimmeringLineView in an overlay

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

staskus committed:
Delay orderState syncing to loaded change for coupons validation

'Coupon Validation' is an artificial payment state shown when order is loading and coupons added to cart.

Visual glitches can happen due to a small delay between syncing state finishing and payment starting.

I want to solve this issue isolating the coupons validation without affecting any other states.

Introduced TotalsViewCouponValidationStateModel which introduces a delay between syncing to loading changes which makes coupons validation appear a bit longer to avoid visual glitches.

staskus committed:
Don't use delayed syncing if card reader is disconnected

staskus commented:
@joshheald @iamgabrielma it's a very custom solution to support what looked to be a relatively simple state. 

I either went the wrong direction, and there's an easier way to do this, or introducing an artificial 'Coupon Validation' state goes too much against the stream of the current structure.

iamgabrielma commented:
> Remove 'Validating Coupons' state from scope

I haven't reviewed/tested the PR yet, just an early note: I'm keen to remove any artificial state if it's not really needed, if I'm not mistaken, AFAIU the introduction of this state steamed from design's understanding of the flow, not from a real necessity to have it.

Does it add any value to the merchant to know at which state of validation are they? Personally I'd say it's close to zero, so perhaps it's also not that valuable to introduce a new artificial payment state to cover it.

staskus commented:
@iamgabrielma thanks for your input, I opened the discussion.
